README.md
# FlashBox - Esp32 serial flasher

[![Watch the flashbox video](https://github.com/Machiel80/esp32-serial-flasher/blob/main/flashbox%20design/video/youtube%20button.png)](https://youtu.be/JwLJvGU8xQY)

## Overview
The FlashBox is an Esp32 (host) to Esp32 (target) serial flasher. 

This concept is especially useful when you need to produce a small quantity of PCB with embedded ESP32 software. The process of designing 
(for example with EasyEDA.com) and producing (for example from jlcpcb.com) PCB’s is relative straightforward. But in small quantities 
(less than 1000 pcs) it’s made without your embedded software on the PCB. It's up to you how to get your software into the ESP32 chip.<br/>
<br/>
This FashBox project is meant to making the flashing process straightforward as well, see the video for a quick demonstration. It makes the flash process quick, easy to execute (press the button), reliable (green LED will turn-on when software is verified) and finally the process is scalable. Scalable because, a FlashBox can operate independently from a computer and is cheap to build, a rough estimation of the out-of-pocket cost is around 30 dollars [^1]. Building more than one FlashBox makes the project scalable and gives you the opportunity to outsource the flashing process.

## Costomize

The software and hardware design of this project is open-source and free to use. Based on your own PCB you need to customize the soft- and hardware to let it fit for your needs. The FlashBox case and lid are 3D printed. These design files are published and ready to customize for your own case. Furthermore, there’s a bill of material (BOM) available with all the electronic components that is used to build the FlashBox. All those components, like display and buttons, are covered by a AliExpress link. For the wiring of the components there’s a hardware wire diagram available to assemble the electronics circuit. When you are designing your own target PCB, make sure that the +5V, GND, IO0, RX, TX and EN are available. For those who would like to use the default ESP32 developer board as target PCB, (the same that is used inside the FlashBox) there is a second wire diagram available.

## Technical
The FlashBox software is build on the 'Serial flasher' component from espressif. This component is a software library for flashing Espressif SoCs from other host microcontroller. Espressif SoCs are normally programmed via serial interface (UART). Port layer for given host microcontroller has to be implemented, if not available.

Your own ESP32 binary must be placed in the `firmware_target` directory and will be converted into C-array during the build process.

Following steps are performed in order to re-program target's memory:

1. Filesystem is initialized and mounted.
2. UART1 through which new binary will be transfered is initialized.
3. Host puts target device into boot mode tries to connect by calling `esp_loader_connect()`.
4. Binary file is opened and its size is acquired, as it has to be known before flashing.
5. Then `esp_loader_flash_start()` is called to enter flashing mode and erase amount of memory to be flashed.
6. `esp_loader_flash_write()` function is called repeatedly until the whole binary image is transfered.

[^1]: See 'flashbox design/components/BOM.xlsx'

## Target firmware preparation outline

1. Add a UART 'version' command to the target firmware that returns the current version.<br/>
   See `sourcecode example target` directory for an implementation example.
2. Compile the target firmware and place the .bin files in the `firmware_target/bin` directory.
3. Add the version number from step 1 and change the TARGET_SOFTWARE_VERSION in `main/main.h`.
4. Update the filenames in the `main/main.h` file.<br/>
   The partition, boot, ota and application filenames should match with the names of your bin files (`firmware_target/bin` directory).<br/>
   Tip: look into the `build/binaries.c` for the correct spelling of the names.<br/>
   Tip: nevert start a bin filename with a number.
5. Compile and flash the FlashBox software

## Hardware connection
Table below shows connection between the FlashBox host and thet target ESP32 devices.

| ESP32 (host) | ESP32 (target)|
|:------------:|:-------------:|
|    IO35      |      TX0      |
|    IO32      |      IO0      |
|    IO33      |     RESET     |
|    IO25      |      RX0      |


## Installation history
git clone https://github.com/espressif/esp-serial-flasher.git<br/>
git clone https://github.com/craftmetrics/esp32-button.git<br/>
git clone https://github.com/olikraus/u8g2.git<br/>
git clone https://github.com/mdvorak/esp-u8g2-hal.git<br/>
<br/>
change (/components/esp32-button/src/button.c) line 132 to 'debounce[idx].inverted = false;'<br/>
change (/components/esp-u8g2-hal/include/u8g2_esp32_hal.h) line 27 to 'I2C_MASTER_FREQ_HZ 250000'<br/>

## 3D-Files
The 3D case design is open-source, see folder `flashbox design`

![3D design impression](https://github.com/Machiel80/esp32-serial-flasher/blob/main/flashbox%20design/3d%20model/20220907_FlashBox_right.png)

## Inspiration
https://github.com/tuanpmt/esp_rpi_flasher<br/>
https://www.youtube.com/watch?v=gd818KGObi0<br/>
https://eprotutorials.com/esp32-gpio-tutorial/<br/>
https://github.com/espressif/esp-serial-flasher<br/>
http://www.lucadentella.it/en/2017/10/30/esp32-25-display-oled-con-u8g2<br/>
https://github.com/lucadentella/esp32-tutorial/tree/master/18_u8g2<br/>
https://github.com/mkfrey/u8g2-hal-esp-idf<br/>
https://github.com/mdvorak/esp-u8g2-hal<br/>
https://www.mischianti.org/images-to-byte-array-online-converter-cpp-arduino<br/>
https://convertio.co/bmp-xbm/<br/>
https://www.esp32.com/viewtopic.php?t=18656<br/>
https://www.youtube.com/watch?v=eLmpKKaQL54<br/>
https://randomnerdtutorials.com/solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header<br/>

## Hardware wire diagram
![flashbox wire diagram with dev board](https://github.com/Machiel80/esp32-serial-flasher/blob/main/flashbox%20design/doc/flashbox%20wire%20diagram.png?raw=true)<br/>
Notice that the IO0 is not available on the default pins. The blue wire in the diagram is a shortcut to IO0.<br/>

![flashbox wire diagram](https://github.com/Machiel80/esp32-serial-flasher/blob/main/flashbox%20design/doc/flashbox%20wire%20diagram%20with%20target.png?raw=true)<br/>
Notice the 10 uF electrolytic capacitor between the EN pin and GND that's used to get the ESP32 into flashing/uploading mode automatically when uploading a new code.<br/>
<br/>



CHANGELOG
Version esp32-serial-flasher

20220906: Removed a lot of custom code, add example source code, update documentation
20220421: Code clean and add documentation
20220420: Initial version



main/main.c
/* 
    name: Esp32 serial flasher
    from: M.Mastenbroek (github.com/Machiel80)
*/

#include <sys/param.h>
#include <string.h>
#include "main.h"
#include "esp_err.h"
#define LOG_LOCAL_LEVEL ESP_LOG_DEBUG // ESP_LOG_VERBOSE
#include "esp_log.h"
#include "driver/uart.h"
#include "driver/gpio.h"
#include "esp32_port.h"
#include "esp_loader.h"
#include "firmware_target.h"
#include <u8g2.h>
#include "u8g2_esp32_hal.h"
#include "button.h"
#include "icons.h"

/* extern "C" {
	void app_main(void);
} */

volatile bool _toggle_validate_slave_software_task_enabled   = true;
volatile bool _toggle_result_validate_slave_software_changed = true;
u8g2_t _u8g2;  // a structure which will contain all the data for one display
enum cookerFirmwareVersion _firmware_version = cfvUNKNOWN;

static void change_request_validate_software(const char* note, enum cookerFirmwareVersion firmware_version_state) {
    const char* TAG = "software changed";
    esp_log_level_set(TAG, ESP_LOG_INFO);

    ESP_LOGD(TAG,"request '%s'",note);

    if(firmware_version_state != _firmware_version ) {
        _firmware_version = firmware_version_state;
        _toggle_result_validate_slave_software_changed = true;
        if(_firmware_version == cfvVALIDATED)
            ESP_LOGD(TAG,"state changed '%s' to VALIDATED",note);
        else if(_firmware_version == cfvAMISS)
            ESP_LOGD(TAG,"state changed '%s' to AMISS",note);
        else
            ESP_LOGD(TAG,"state changed '%s' to UNKNOWN",note);
    }
}

static void slave_software_validate_task(void *arg) {
    static const char* TAG = "validate software";
    static bool task_restart = false;
    esp_log_level_set(TAG, ESP_LOG_INFO);
    
    unsigned long REQUEST_FOR_VERSION_INTERVAL = 1000; // ms example: 1000 = 1 sec.
    unsigned long RESPONSE_FOR_VERSION_TIMEOUT =  200; // ms example: 1000 = 1 sec.
	unsigned long lastVersionRequestTime = 0;
    unsigned long lastVersionResponseTime = 0;

    bool version_request_sent = false;
    bool timeout_timer_active = false;

    /* Configure parameters of an UART driver,
     * communication pins and install the driver */
    uart_config_t uart_config = {
        .baud_rate = VALIDATE_UART_BAUD_RATE,
        .data_bits = UART_DATA_8_BITS,
        .parity    = UART_PARITY_DISABLE,
        .stop_bits = UART_STOP_BITS_1,
        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
        .source_clk = UART_SCLK_APB,
    };
    
    int intr_alloc_flags = 0;

#if CONFIG_UART_ISR_IN_IRAM
    intr_alloc_flags = ESP_INTR_FLAG_IRAM;
#endif

    ESP_ERROR_CHECK(uart_driver_install(VALIDATE_UART_PORT_NUM, BUF_SIZE * 2, 0, 0, NULL, intr_alloc_flags));
    ESP_ERROR_CHECK(uart_param_config(VALIDATE_UART_PORT_NUM, &uart_config));
    ESP_ERROR_CHECK(uart_set_pin(VALIDATE_UART_PORT_NUM, VALIDATE_TEST_TXD, VALIDATE_TEST_RXD, VALIDATE_TEST_RTS, VALIDATE_TEST_CTS));

    // Configure a temporary buffer for the incoming data
    uint8_t *data = (uint8_t *) malloc(BUF_SIZE);

    ESP_LOGI(TAG,"(I) task started");
    ESP_LOGD(TAG,"(D) task started");

    while (true) {
        if(_toggle_validate_slave_software_task_enabled == true) {
            if(task_restart == true) {
                task_restart = false;
                ESP_LOGD(TAG,"task restarted");
                ESP_ERROR_CHECK(uart_driver_install(VALIDATE_UART_PORT_NUM, BUF_SIZE * 2, 0, 0, NULL, intr_alloc_flags));
                ESP_ERROR_CHECK(uart_param_config(VALIDATE_UART_PORT_NUM, &uart_config));
                ESP_ERROR_CHECK(uart_set_pin(VALIDATE_UART_PORT_NUM, VALIDATE_TEST_TXD, VALIDATE_TEST_RXD, VALIDATE_TEST_RTS, VALIDATE_TEST_CTS));

                ESP_LOGI(TAG, "reset slave");
                esp_loader_reset_target();
                vTaskDelay(100 / portTICK_PERIOD_MS);
            }

            // Read data from the UART
            int len = uart_read_bytes(VALIDATE_UART_PORT_NUM, data, BUF_SIZE, 20 / portTICK_RATE_MS);

            if(len > 0) {
                if(version_request_sent == true) {
                    version_request_sent = false;
                    if (memcmp ( data, SAFETYCOOK_SOFTWARE_VERSION, strlen(SAFETYCOOK_SOFTWARE_VERSION) ) == 0) {
                        change_request_validate_software("equal versions", cfvVALIDATED);
                    }else {
                        change_request_validate_software("unknown version", cfvAMISS);
                    }
                    timeout_timer_active = false;
                } else if(len > 5) { // slave is sending UART information, ignore input and wait until 
                    lastVersionRequestTime = esp_timer_get_time() / 1000ULL; // restart timer
                    change_request_validate_software("noise, version unknown", cfvUNKNOWN);
                }
                // ESP_LOGI(TAG,"%.*s",len, data);
                // printf("%.*s",len, data);
            } 

            if((esp_timer_get_time() / 1000ULL) - lastVersionRequestTime >= REQUEST_FOR_VERSION_INTERVAL) {
                lastVersionRequestTime += REQUEST_FOR_VERSION_INTERVAL; // reset timer
    
                if(version_request_sent == false) {
                    version_request_sent = true;
                    ESP_LOGD(TAG,"send UART request for version information");
                    // Write data back to the UART
                    uart_write_bytes(VALIDATE_UART_PORT_NUM, "version\n",strlen("version\n"));
                    lastVersionResponseTime = esp_timer_get_time() / 1000ULL; // restart timer
                    timeout_timer_active = true;
                }else {
                    change_request_validate_software("time-out (4 seconds)", cfvAMISS);
                    timeout_timer_active = false;
                    version_request_sent = false;
                }
            } // timer, request for version 

            if(timeout_timer_active == true && (esp_timer_get_time() / 1000ULL) - lastVersionResponseTime >= RESPONSE_FOR_VERSION_TIMEOUT) {
                timeout_timer_active = false;
                change_request_validate_software("time-out", cfvAMISS);
            } // timer, response for version 
        }else {
            
            if(task_restart == false) {
                task_restart = true;
                ESP_LOGD(TAG,"task pauze, UART driver turned-off");
                uart_driver_delete(VALIDATE_UART_PORT_NUM);
                change_request_validate_software("task turned off",cfvUNKNOWN);
            }
            vTaskDelay(200 / portTICK_PERIOD_MS);
        }
    } // endless loop
}

void set_icon(enum icon icon_name) {
    u8g2_ClearBuffer(&_u8g2);
    switch (icon_name) {
      case completed:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_completed);
        break;
      case failed:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_failed);
        break;
      case progress_1_3:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_progress_1_3);
        break;
      case progress_2_3:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_progress_2_3);
        break;
      case progress_3_3:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_progress_3_3);
        break;
      case open:
        u8g2_DrawXBM(&_u8g2, 32, 0, 64, 64, icon_open);
        u8g2_SetFont(&_u8g2, u8g2_font_6x13_tf);
        u8g2_DrawStr(&_u8g2, 40, 36, SAFETYCOOK_SOFTWARE_VERSION);
        break;    
    }
    u8g2_SendBuffer(&_u8g2);
}

void app_main(void) {
    printf("Start ...\n" );
    static const char* TAG = "main";
    esp_log_level_set(TAG, ESP_LOG_DEBUG);

    gpio_config_t io_conf; // Fix issue with GPIO 14 (RED LED) didn't work well
    io_conf.intr_type = GPIO_INTR_DISABLE; // disable interrupt
    io_conf.mode = GPIO_MODE_OUTPUT;       // set as output mode
    io_conf.pin_bit_mask = GPIO_OUTPUT_PIN_SEL; // bit mask of the pins that you want to set, e.g.GPIO14/27
    io_conf.pull_down_en = (gpio_pulldown_t) 0; // disable pull-down mode
    io_conf.pull_up_en = (gpio_pullup_t)0;      // disable pull-up mode
    gpio_config(&io_conf);   // configure GPIO with the given settings

    gpio_set_direction((gpio_num_t) SSD1306_PWR, GPIO_MODE_OUTPUT);
    gpio_set_direction((gpio_num_t) SSD1306_GND, GPIO_MODE_OUTPUT);
    gpio_set_direction((gpio_num_t) RED_LED_PWR, GPIO_MODE_OUTPUT);
    gpio_set_direction((gpio_num_t) RED_LED_GND, GPIO_MODE_OUTPUT);
    gpio_set_direction((gpio_num_t) GREEN_LED_PWR, GPIO_MODE_OUTPUT);
    gpio_set_direction((gpio_num_t) GREEN_LED_GND, GPIO_MODE_OUTPUT);
    gpio_set_level((gpio_num_t) SSD1306_PWR, 1); // +3.3V
    gpio_set_level((gpio_num_t) SSD1306_GND, 0); // GND
    gpio_set_level((gpio_num_t) RED_LED_PWR, 0);
    gpio_set_level((gpio_num_t) RED_LED_GND, 0);
    gpio_set_level((gpio_num_t) GREEN_LED_PWR, 0);
    gpio_set_level((gpio_num_t) GREEN_LED_GND, 0); // GND

    u8g2_esp32_hal_t u8g2_esp32_hal = U8G2_ESP32_HAL_DEFAULT;
    u8g2_esp32_hal.sda = (gpio_num_t) SSD1306_SDA;;  // data for I²C
    u8g2_esp32_hal.scl = (gpio_num_t) SSD1306_SCL;;  // clock for I²C

    // u8g2_esp32_hal.clk = (gpio_num_t) (-1);
    // u8g2_esp32_hal.mosi = (gpio_num_t) (-1);
    // u8g2_esp32_hal.cs = (gpio_num_t) (-1);
    // u8g2_esp32_hal.reset = (gpio_num_t) (-1);
    // u8g2_esp32_hal.dc = (gpio_num_t) (-1);
    // u8g2_esp32_hal.i2c_num = (gpio_num_t) (-1);
    // u8g2_esp32_hal.i2c_clk_speed = (gpio_num_t) (-1);

    u8g2_esp32_hal_init(u8g2_esp32_hal);

    u8g2_Setup_ssd1306_i2c_128x64_noname_f(
          &_u8g2
        , U8G2_R0
        , u8g2_esp32_i2c_byte_cb
        , u8g2_esp32_gpio_and_delay_cb);  // init u8g2 structure
    
    // Note that address '0x78' is the I2C address already shifted left to include the read/write flag. 
    // Instead of supplying the address 0x3C which would be 0011 1100 supply 0x78 which would be 0111 1000.
    u8x8_SetI2CAddress(&_u8g2.u8x8, 0x78); 

    ESP_LOGI(TAG, "u8g2 InitDisplay");
    u8g2_InitDisplay(&_u8g2);  // send init sequence to the display, display is in, sleep mode after this
    ESP_LOGI(TAG, "u8g2 SetPowerSave");
    u8g2_SetPowerSave(&_u8g2, 0);  // wake up display
  
    u8g2_ClearBuffer(&_u8g2);
    u8g2_SetFont(&_u8g2, u8g2_font_tallpixelextended_tf);
    u8g2_DrawStr(&_u8g2, 31, 11, "SafetyCook");
    u8g2_SetFont(&_u8g2, u8g2_font_fub20_tn);
    u8g2_DrawStr(&_u8g2, 2, 41, SAFETYCOOK_SOFTWARE_VERSION);
    u8g2_SetFont(&_u8g2, u8g2_font_6x13_tf);
    u8g2_DrawStr(&_u8g2, 2, 63, "FlashBox:");
    u8g2_DrawStr(&_u8g2, 60, 63, FLASHER_SOFTWARE_VERSION);

    u8g2_SendBuffer(&_u8g2);
    vTaskDelay(5000 / portTICK_PERIOD_MS);

    target_binaries_t bin; // flash binary configuration

    const loader_esp32_config_t config = {
        .baud_rate = VALIDATE_UART_BAUD_RATE, // 115200
        .uart_port = UART_NUM_1,
        .uart_rx_pin =  TARGET_RX_TX, 
        .uart_tx_pin =  TARGET_TX_RX, 
        .reset_trigger_pin =  TARGET_RESET, 
        .gpio0_trigger_pin =  TARGET_IO0,
     // .rx_buffer_size = 0,    /*!< Set to zero for default RX buffer size */
     // .tx_buffer_size = 0,    /*!< Set to zero for default TX buffer size */        
     // .queue_size = 0,        /*!< Set to zero for default UART queue size */
     // .uart_queue = NULL      /*!< Set to NULL, if UART queue handle is not necessary. Otherwise, it will be assigned here */
    };

    button_event_t ev;
    QueueHandle_t button_events = button_init(PIN_BIT(BUTTON_1));
    xTaskCreate(slave_software_validate_task, "software validate task", VALIDATE_TASK_STACK_SIZE, NULL, 10, NULL);

    while (true) { // endless loop

        if(_toggle_result_validate_slave_software_changed == true) {
            _toggle_result_validate_slave_software_changed = false;

            ESP_LOGI(TAG, "validate slave software, state changed");
            if(_firmware_version == cfvVALIDATED) {
                gpio_set_level((gpio_num_t) GREEN_LED_PWR, 1);
                set_icon(completed);
                ESP_LOGI(TAG, "slave software validated");
            } else if(_firmware_version == cfvAMISS) {
                gpio_set_level((gpio_num_t) GREEN_LED_PWR, 0);
                set_icon(open);
                ESP_LOGI(TAG, "slave software amiss");
            } else { // cfvUNKNOWN
                gpio_set_level((gpio_num_t) GREEN_LED_PWR, 0);
                // set_icon(open);
                ESP_LOGI(TAG, "slave software unknown");
            }
        }

        if (xQueueReceive(button_events, &ev, 1000/portTICK_PERIOD_MS)) {
            if ((ev.pin == BUTTON_1) && (ev.event == BUTTON_DOWN)) {

                ESP_LOGI(TAG, "button pressed");
                set_icon(progress_1_3);
                gpio_set_level((gpio_num_t) RED_LED_PWR, 1);
                gpio_set_level((gpio_num_t) GREEN_LED_PWR, 0);
                _toggle_validate_slave_software_task_enabled = false;
                vTaskDelay(100 / portTICK_PERIOD_MS);

                if (loader_port_esp32_init(&config) != ESP_LOADER_SUCCESS) {
                    ESP_LOGE(TAG, "serial initialization failed");
                    set_icon(failed);
                } else {
                    ESP_LOGI(TAG, "serial initialization completed");
                    if (connect_to_target(HIGHER_BAUDRATE) != ESP_LOADER_SUCCESS) {
                        ESP_LOGE(TAG, "connect to target failed");
                        set_icon(failed);
                    } else {
                        ESP_LOGI(TAG, "connect to target completed");
                        get_binaries(esp_loader_get_target(), &bin);
                        if (flash_binary(bin.boot.data, bin.boot.size, bin.boot.addr) != ESP_LOADER_SUCCESS) {
                            ESP_LOGE(TAG, "flash binary 'boot' failed");
                            set_icon(failed);
                        } else {
                            ESP_LOGI(TAG, "flash binary 'boot' completed");
                            
                            if(flash_binary(bin.part.data, bin.part.size, bin.part.addr) != ESP_LOADER_SUCCESS) {
                                ESP_LOGE(TAG, "flash binary 'part' failed");
                                set_icon(failed);
                            } else {
                                ESP_LOGI(TAG, "flash binary 'part' completed");
                                    set_icon(progress_2_3);
                                if(flash_binary(bin.app.data,  bin.app.size,  bin.app.addr) != ESP_LOADER_SUCCESS) {
                                    ESP_LOGE(TAG, "flash binary 'app' failed");
                                    set_icon(failed);
                                } else {
                                    ESP_LOGI(TAG, "flash binary 'app' completed");
                                    set_icon(progress_3_3);
#ifdef SAFETYCOOK_ROM
                                    if(flash_binary(bin.ota.data,  bin.ota.size,  bin.ota.addr) != ESP_LOADER_SUCCESS) {
                                        ESP_LOGE(TAG, "flash binary 'ota' failed");
                                        set_icon(failed);
                                    } else {    
                                        ESP_LOGI(TAG, "flash binary 'ota' completed");                    
                                    } // ota completed
#endif
                                } // app completed
                            } // part completed
                        } // boot completed
                    } // connect to slave completed
                } // serial initialization completed

                ESP_LOGI(TAG, "uart driver delete");
                uart_driver_delete(config.uart_port);
                xQueueReset(button_events); // remove button presses 
                gpio_set_level((gpio_num_t) RED_LED_PWR, 0);
                _toggle_validate_slave_software_task_enabled = true;
                
            } // button 1 pressed
        } // some button events
    } // endless loop
} // app main

firmware_target/firmware_target.c
/* Copyright 2020 Espressif Systems (Shanghai) PTE LTD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdio.h>
#include <string.h>
#include <sys/param.h>
#include "serial_io.h"
#include "esp_loader.h"
// #define LOG_LOCAL_LEVEL ESP_LOG_INFO // ESP_LOG_DEBUG, ESP_LOG_VERBOSE
#include "esp_log.h"
#include "firmware_target.h"
#include "main.h"

#define BOOTLOADER_ADDRESS  0x1000
#define PARTITION_ADDRESS   0x8000
#define OTA_ADDRESS         0xe000   // OTA_ADDRESS: See: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html#ota-data-partition
#define APPLICATION_ADDRESS 0x10000  

extern const uint8_t  BOOTLOADER_BIN[];
extern const uint32_t BOOTLOADER_BIN_SIZE;
extern const uint8_t  PARTITION_TABLE_BIN[];
extern const uint32_t PARTITION_TABLE_BIN_SIZE;
extern const uint8_t  APPLICATION_BIN[];
extern const uint32_t APPLICATION_BIN_SIZE;

#ifdef OTA_ROM
extern const uint8_t  OTA_BIN[];
extern const uint32_t OTA_BIN_SIZE;
#endif

void get_binaries(target_chip_t target, target_binaries_t *bins) {
    
    bins->boot.data = BOOTLOADER_BIN;
    bins->boot.size = BOOTLOADER_BIN_SIZE;
    bins->boot.addr = BOOTLOADER_ADDRESS;
    bins->part.data = PARTITION_TABLE_BIN;
    bins->part.size = PARTITION_TABLE_BIN_SIZE;
    bins->part.addr = PARTITION_ADDRESS;
    bins->app.data  = APPLICATION_BIN;
    bins->app.size  = APPLICATION_BIN_SIZE;
    bins->app.addr  = APPLICATION_ADDRESS;
#ifdef OTA_ROM
    bins->ota.data  = OTA_BIN;
    bins->ota.size  = OTA_BIN_SIZE;
    bins->ota.addr  = OTA_ADDRESS;
#endif
}

esp_loader_error_t connect_to_target(uint32_t higher_baudrate) {
    static const char* TAG = "connect to target";
    esp_loader_connect_args_t connect_config = ESP_LOADER_CONNECT_DEFAULT();

    esp_loader_error_t err = esp_loader_connect(&connect_config);
    if (err != ESP_LOADER_SUCCESS) {
        ESP_LOGD(TAG,"cannot connect to target. Error: %u", err);
        return err;
    }
     ESP_LOGI(TAG,"connected to target");

    if (higher_baudrate && esp_loader_get_target() != ESP8266_CHIP) {
        err = esp_loader_change_baudrate(higher_baudrate);
        if (err == ESP_LOADER_ERROR_UNSUPPORTED_FUNC) {
            ESP_LOGD(TAG,"ESP8266 does not support change baudrate command");
            return err;
        } else if (err != ESP_LOADER_SUCCESS) {
            ESP_LOGD(TAG,"unable to change baud rate on target");
            return err;
        } else {
            err = loader_port_change_baudrate(higher_baudrate);
            if (err != ESP_LOADER_SUCCESS) {
                ESP_LOGD(TAG,"unable to change baud rate");
                return err;
            }
            ESP_LOGI(TAG,"baudrate changed");
        }
    }
    return ESP_LOADER_SUCCESS;
}

esp_loader_error_t flash_binary(const uint8_t *bin, size_t size, size_t address) {
    static const char* TAG = "flash binary";
    esp_loader_error_t err;
    static uint8_t payload[1024];
    const uint8_t *bin_addr = bin;

    ESP_LOGI(TAG,"erasing flash (this may take a while)...");
    err = esp_loader_flash_start(address, size, sizeof(payload));
    if (err != ESP_LOADER_SUCCESS) {
        ESP_LOGD(TAG,"erasing flash failed with error %d.\n", err);
        return err;
    }
    ESP_LOGI(TAG,"start programming");

    size_t binary_size = size;
    size_t written = 0;

    while (size > 0) {
        size_t to_read = MIN(size, sizeof(payload));
        memcpy(payload, bin_addr, to_read);

        err = esp_loader_flash_write(payload, to_read);
        if (err != ESP_LOADER_SUCCESS) {
            ESP_LOGD(TAG,"packet could not be written! error %d", err);
            return err;
        }

        size -= to_read;
        bin_addr += to_read;
        written += to_read;

        int progress = (int)(((float)written / binary_size) * 100);
        ESP_LOGD(TAG,"progress: %d %%", progress);
        fflush(stdout);
    };

    ESP_LOGI(TAG,"finished programming");

#if MD5_ENABLED
    err = esp_loader_flash_verify();
    if (err == ESP_LOADER_ERROR_UNSUPPORTED_FUNC) {
        ESP_LOGD(TAG,"ESP8266 does not support flash verify command");
        return err;
    } else if (err != ESP_LOADER_SUCCESS) {
        ESP_LOGD(TAG,"MD5 does not match. err: %d", err);
        return err;
    }
    ESP_LOGI(TAG,"flash verified");
#endif

    return ESP_LOADER_SUCCESS;
}

main/main.h
// #define OTA_ROM

#define BOOTLOADER_BIN           bootloader_bin
#define BOOTLOADER_BIN_SIZE      bootloader_bin_size
#define PARTITION_TABLE_BIN      partition_table_bin
#define PARTITION_TABLE_BIN_SIZE partition_table_bin_size
#define APPLICATION_BIN          blink_bin
#define APPLICATION_BIN_SIZE     blink_bin_size
#define OTA_BIN                  ota_data_initial_bin
#define OTA_BIN_SIZE             ota_data_initial_bin_size

// Tip -> update (add empty line and save) /main/CMakeLists.txt to rebuild the .bin files into 'build/binaries.c'

#define TARGET_NAME              "Esp32Blink"
#define FLASHER_SOFTWARE_VERSION "20220907"
#define TARGET_SOFTWARE_VERSION  "20220906"

#define HIGHER_BAUDRATE 230400

#define TARGET_RX_TX  35 // yellow wire
#define TARGET_IO0    32 // blue wire
#define TARGET_RESET  33 // orange wire
#define TARGET_TX_RX  25 // green wire
#define RED_LED_PWR   27 // red wire
#define RED_LED_GND   14 // black wire

#define GREEN_LED_PWR 23 // red wire
#define GREEN_LED_GND 22 // black wire
#define SSD1306_PWR   21 // red wire
#define SSD1306_GND   19 // black wire
#define SSD1306_SDA   18 // yellow wire
#define SSD1306_SCL   05 // white wire
#define BUTTON_1      15 // yellow wire

#define GPIO_OUTPUT_PIN_SEL  ((1ULL<<RED_LED_PWR) | (1ULL<<RED_LED_GND) | (1ULL<<GREEN_LED_PWR)  | (1ULL<<GREEN_LED_GND) )

#define BUF_SIZE (1024)
#define VALIDATE_UART_BAUD_RATE     115200
#define VALIDATE_UART_PORT_NUM      2
#define VALIDATE_TEST_RXD TARGET_RX_TX
#define VALIDATE_TEST_TXD TARGET_TX_RX
#define VALIDATE_TEST_RTS (UART_PIN_NO_CHANGE)
#define VALIDATE_TEST_CTS (UART_PIN_NO_CHANGE)
#define VALIDATE_TASK_STACK_SIZE    2048

#define U8G2_ESP32_HAL_DEFAULT_2  {        \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    (gpio_num_t) U8G2_ESP32_HAL_UNDEFINED, \
    I2C_MASTER_NUM,                        \
    I2C_MASTER_FREQ_HZ   }


enum cookerFirmwareVersion {
  cfvVALIDATED,
  cfvAMISS,
  cfvUNKNOWN
};

enum icon {
    completed
   ,failed
   ,progress_1_3
   ,progress_2_3
   ,progress_3_3
   ,open
};
